From a6366f1061694388e6354bc88867d8e1ef065b7d Mon Sep 17 00:00:00 2001
From: Sagar Ghuge <sagar.ghuge@intel.com>
Date: Thu, 20 May 2021 23:48:47 -0700
Subject: [PATCH 18/25] intel/compiler: Add Wa_14014063774 for xe_hpg

If there are multiple SLM load SBID that are live, then the immeidate
mask for sync.allwr should include all the corresponding SBID bits.

Signed-off-by: Sagar Ghuge <sagar.ghuge@intel.com>
---
 src/intel/compiler/brw_fs_nir.cpp | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/intel/compiler/brw_fs_nir.cpp b/src/intel/compiler/brw_fs_nir.cpp
index 6a3bb21f3c5..b06cc940f7f 100644
--- a/src/intel/compiler/brw_fs_nir.cpp
+++ b/src/intel/compiler/brw_fs_nir.cpp
@@ -4526,6 +4526,18 @@ fs_visitor::nir_emit_intrinsic(const fs_builder &bld, nir_intrinsic_instr *instr
 
          if (slm_fence) {
             assert(opcode == SHADER_OPCODE_MEMORY_FENCE);
+            if (intel_device_info_is_dg2(devinfo)) {
+               /* Wa_14014063774
+                *
+                * If there are multiple SLM load SBID that are live, then the
+                * immeidate mask for sync.allwr should include all the
+                * corresponding SBID bits.
+                *
+                */
+               ubld.exec_all().group(1, 0).emit(
+                  BRW_OPCODE_SYNC, ubld.null_reg_ud(),
+                  brw_imm_ud(TGL_SYNC_ALLWR));
+            }
             fence_regs[fence_regs_count++] =
                emit_fence(ubld, opcode, GFX12_SFID_SLM,
                           true /* commit_enable */,
-- 
2.34.1

